kv/kvserver: TestLearnerReplicateQueueRace failed

kv/kvserver.TestLearnerReplicateQueueRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8281964?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8281964?buildTab=artifacts#/) on master @ [033a395ae83f8175eee609bd4918a31cf807fc9b](https://github.com/cockroachdb/cockroach/commits/033a395ae83f8175eee609bd4918a31cf807fc9b):

Fatal error:

```
panic: test timed out after 59m55s

```
Stack: 

```
goroutine 178283 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* google.golang.org/grpc.(*Server).serveStreams(0x400543d860, {0x6897bc0?, 0x4006e8d380})
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:975 +0x148
* google.golang.org/grpc.(*Server).handleRawConn.func1()
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:891 +0x44
* created by google.golang.org/grpc.(*Server).handleRawConn
* 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:890 +0x184
* 
* goroutine 169208 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x684c6f8, 0x40073e4360})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x684c688?, 0x40001b2000?}, {{0x400035fd60?, 0x684c6f8?, 0x0?}}, 0x400d5ef798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40068441e0?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 168669 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:419 +0x224
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:392 +0x60
* 
* goroutine 168785 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x684c6f8, 0x400bcf6f90})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x684c688?, 0x40001b2000?}, {{0x400035fd60?, 0x400063a601?, 0x0?}}, 0x4001d8e798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4001d8e758?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 169221 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x684c6f8, 0x40073e5ce0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x684c688?, 0x40001b2000?}, {{0x400035fd60?, 0x400d727680?, 0x10160e0?}}, 0x40052b0798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40052b0758?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 168667 [select]:
* github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:419 +0x224
* created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs
* 	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:392 +0x60
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerReplicateQueueRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23353

kv/kvserver.TestLearnerReplicateQueueRace [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8366381?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8366381?buildTab=artifacts#/) on master @ [38d207af32dbc568982f95ac6eb8c66b0bfc03cf](https://github.com/cockroachdb/cockroach/commits/38d207af32dbc568982f95ac6eb8c66b0bfc03cf):

Fatal error:

```
panic: test timed out after 59m55s

```
Stack: 

```
goroutine 205857 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x68b0c80, 0x4006aba150})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x68b0c10?, 0x40001b2000?}, {{0x4000642300?, 0x68b0c80?, 0x0?}}, 0x4003eeff98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4003eeff58?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 196994 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x68b0c80, 0x400cdc2330})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x68b0c10?, 0x40001b2000?}, {{0x4000642300?, 0x68b0c80?, 0x0?}}, 0x4003fde798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4003fde758?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 196947 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x68b0c80, 0x4006aba8d0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x68b0c10?, 0x40001b2000?}, {{0x4000642300?, 0x4008f92f90?, 0x0?}}, 0x4003ef4798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x400d5fb2a0?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 196695 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x68b0c80, 0x4011ea00f0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x68b0c10?, 0x40001b2000?}, {{0x4000642300?, 0x68b0c01?, 0x0?}}, 0x400bc30f98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x400bc30f58?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
* goroutine 196743 [chan receive, 3 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x68b0c80, 0x4011ea1110})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:324 +0x80
* runtime/pprof.Do({0x68b0c10?, 0x40001b2000?}, {{0x4000642300?, 0x100004008bcefe0?, 0x10196d0?}}, 0x40084a8f98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40084a8f58?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:322 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:314 +0xf0
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerReplicateQueueRace.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -46,6 +46,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
 )
@@ -1269,9 +1270,10 @@ func TestLearnerReplicateQueueRace(t *testing.T) {
 		// leaving the 2 voters.
 		startKey := tc.ScratchRange(t)
 		desc, err := tc.RemoveVoters(startKey, tc.Target(2))
-		require.NoError(t, err)
-		require.Len(t, desc.Replicas().VoterDescriptors(), 2)
-		require.Len(t, desc.Replicas().LearnerDescriptors(), 0)
+		// NB: don't fatal on this goroutine, as we can't recover cleanly.
+		assert.NoError(t, err)
+		assert.Len(t, desc.Replicas().VoterDescriptors(), 2)
+		assert.Len(t, desc.Replicas().LearnerDescriptors(), 0)
 		return false
 	}
 	tc = testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
@@ -1300,10 +1302,8 @@ func TestLearnerReplicateQueueRace(t *testing.T) {
 			if err != nil {
 				return err
 			}
-			if !strings.Contains(processErr.Error(), `descriptor changed`) {
-				// NB: errors.Wrapf(nil, ...) returns nil.
-				// nolint:errwrap
-				return errors.Errorf(`expected "descriptor changed" error got: %+v`, processErr)
+			if processErr == nil || !strings.Contains(processErr.Error(), `descriptor changed`) {
+				return errors.Wrap(processErr, `expected "descriptor changed" error got: %+v`)
 			}
 			formattedTrace := trace.String()
 			expectedMessages := []string{

